### ISSUE TITLE
BrianObjectException Error with User-provided functions

### ISSUE THREAD

**Issue by JingjinW** [brian-team/brian2#1018]
Hi,

Not sure where I should post my question, I'll post here for now. If there is another site I should go to please let me know.

It doesn't really matter what model I am using, the user provided equation would work outside of model equation, but does not work inside, even if it is not in the calculation really. For simplicity, I will use this  model (https://brian2.readthedocs.io/en/stable/examples/COBAHH.html) to present the problem. 

The function that doesn't work is zeta(v).

```
from brian2 import *

# Parameters
area = 20000*umetre**2
Cm = (1*ufarad*cm**-2) * area
gl = (5e-5*siemens*cm**-2) * area

El = -60*mV
EK = -90*mV
ENa = 50*mV
g_na = (100*msiemens*cm**-2) * area
g_kd = (30*msiemens*cm**-2) * area
VT = -63*mV
# Time constants
taue = 5*ms
taui = 10*ms
# Reversal potentials
Ee = 0*mV
Ei = -80*mV
we = 6*nS  # excitatory synaptic weight
wi = 67*nS  # inhibitory synaptic weight


@check_units(v=1, result=1)
def zeta(v):
    return -1.5-(1./(1.+exp((v+30.)/5.))) 

# The model
eqs = '''
dv/dt = (gl*(El-v)+ge*(Ee-v)+gi*(Ei-v)-
         g_na*(m*m*m)*h*(v-ENa)-
         g_kd*(n*n*n*n)*(v-EK))/Cm : volt
dm/dt = alpha_m*(1-m)-beta_m*m : 1
dn/dt = alpha_n*(1-n)-beta_n*n : 1
dh/dt = alpha_h*(1-h)-beta_h*h : 1
dge/dt = -ge*(1./taue) : siemens
dgi/dt = -gi*(1./taui) : siemens
alpha_m = 0.32*(mV**-1)*(13*mV-v+VT)/
         (exp((13*mV-v+VT)/(4*mV))-1.)/ms : Hz
beta_m = 0.28*(mV**-1)*(v-VT-40*mV)/
        (exp((v-VT-40*mV)/(5*mV))-1)/ms : Hz
alpha_h = 0.128*exp((17*mV-v+VT)/(18*mV))/ms : Hz
beta_h = 4./(1+exp((40*mV-v+VT)/(5*mV)))/ms : Hz
alpha_n = 0.032*(mV**-1)*(15*mV-v+VT)/
         (exp((15*mV-v+VT)/(5*mV))-1.)/ms : Hz
beta_n = .5*exp((10*mV-v+VT)/(40*mV))/ms : Hz
zetav = zeta(v/mV):1

'''

Ecells = NeuronGroup(1, eqs, threshold='v>0*mV',refractory='v > 0*mV',method='rk4')
run(3*ms)

```

in the eqs, if you comment out zetav line, it works. With it, you get 

```
  File "/anaconda3/envs/my-brian2-env/lib/python3.6/site-packages/brian2/core/network.py", line 843, in before_run
    raise brian_object_exception("An error occurred when preparing an object.", obj, ex)

  File "<string>", line unknown
BrianObjectException
```

But outside you could plot zeta(v)

```
v= linspace(-80,80,1001)*mV

zetav = zeta(v/mV)
plot(v/mV,zetav)
```

 I have no idea what has gone wrong. Thanks!

**Comment by mstimberg** [comment]
Hi,

> Not sure where I should post my question, I'll post here for now. If there is another site I should go to please let me know.

in general the best place for anything that is a question (i.e. you are not sure whether it actually is a bug/limitation in Brian) is the [support mailing list](http://groups.google.com/group/briansupport) (mail: briansupport@googlegroups.com)

Back to your issue. First of all, in your case, you shouldn't use a user-defined function in the first place because you can express everything in the equations itself. The disadvantage of a user-defined function is that you provide Python code, therefore you can only run your simulation with Python (`'numpy'`) mode, making it impossible to benefit from the best performance (see [the documentation](https://brian2.readthedocs.io/en/stable/user/computation.html) for more details). Therefore, instead of defining the `zeta` function, use
```Python
zetav = -1.5-(1./(1.+exp((v/mV+30.)/5.)))  : 1
```
as part of the equations.

But disregarding this for a moment, the problem with your current code is the following: when you define your own function, Brian doesn't know what kind of function it is. It might be a function that just calculates stuff based on its input (as in your case), or it might be a function that takes external input or has an internal state like `rand()`, which returns a different number each time you call it. The latter kind of function is problematic when you call it in what we call a subexpression, i.e. a line like `zetav = ...`. Why? Because we substitute the right-hand-side of this expression wherever it is used. This means that e.g. the value that a monitor records would not be the same as the value that was actually used. To deal with this problem, Brian will raise an error and force you to use the `(constant over dt)` flag for stateful functions (and by default, it just assumes your function is stateful), which means that the function will be only evaluated once for a time step. This is actually the error that you get for your code:
```
An error occurred when preparing an object. SyntaxError: The subexpression 'zetav' refers to a stateful function (e.g. rand()). Such expressions should only be evaluated once per timestep, add the 'constant over dt'flag.
(See above for original error message and traceback.)
```
You did not actually see this error, because you were running your code from ipython or a jupyter notebook and ran into issue #964 – this issue has already been fixed in Brian's development version but we did not have a release since then.

So to sum up:
* If possible, do not use a user-defined function but instead express everything as part of the equations
* If this is not possible, you'll have to mark your function as "stateless" by setting its stateless attribute to `True` (in your code, after defining your function, simply write `zeta.stateless = True`). But this will only work in "Python mode"; to run code with Cython/C++, you'll have to implement your function in the respective target language (see [the documentation](https://brian2.readthedocs.io/en/stable/advanced/functions.html#other-code-generation-targets)).

**Comment by JingjinW** [comment]
Hi Marcel,

Thanks for the reply. Next time I'll use the google support group. For this issue only I'll continue here.

I use the second way and add zeta.stateless = True. It works, until I try to put it in a state variable, say dv/dt here. If dv/dt has zeta(v) in it, I have a NotImplementedError. 

Which means, when only add the line 'zeta.stateless = True' after initialization of the function zeta(v) it runs:
```
@check_units(v=1, result=1)
def zeta(v):
    return -1.5-(1./(1.+exp((v+30.)/5.))) 
zeta.stateless = True  # stateless 

# The model
eqs = '''
dv/dt = (gl*(El-v)+ge*(Ee-v)+gi*(Ei-v)-
         g_na*(m*m*m)*h*(v-ENa)-
         g_kd*(n*n*n*n)*(v-EK))/Cm : volt
dm/dt = alpha_m*(1-m)-beta_m*m : 1
dn/dt = alpha_n*(1-n)-beta_n*n : 1
dh/dt = alpha_h*(1-h)-beta_h*h : 1
dge/dt = -ge*(1./taue) : siemens
dgi/dt = -gi*(1./taui) : siemens
alpha_m = 0.32*(mV**-1)*(13*mV-v+VT)/
         (exp((13*mV-v+VT)/(4*mV))-1.)/ms : Hz
beta_m = 0.28*(mV**-1)*(v-VT-40*mV)/
        (exp((v-VT-40*mV)/(5*mV))-1)/ms : Hz
alpha_h = 0.128*exp((17*mV-v+VT)/(18*mV))/ms : Hz
beta_h = 4./(1+exp((40*mV-v+VT)/(5*mV)))/ms : Hz
alpha_n = 0.032*(mV**-1)*(15*mV-v+VT)/
         (exp((15*mV-v+VT)/(5*mV))-1.)/ms : Hz
beta_n = .5*exp((10*mV-v+VT)/(40*mV))/ms : Hz
zetav = zeta(v/mV):1
'''
```

But if I put zeta(v) in dv/dt it doesn't.
```
@check_units(v=1, result=1)
def zeta(v):
    return -1.5-(1./(1.+exp((v+30.)/5.))) 
zeta.stateless = True  # stateless 

# The model
eqs = '''
dv/dt = (gl*(El-v)+ge*(Ee-v)+gi*(Ei-v)-
         g_na*(m*m*m)*h*(v-ENa)-
        zetav * g_kd*(n*n*n*n)*(v-EK))/Cm : volt
dm/dt = alpha_m*(1-m)-beta_m*m : 1
dn/dt = alpha_n*(1-n)-beta_n*n : 1
dh/dt = alpha_h*(1-h)-beta_h*h : 1
dge/dt = -ge*(1./taue) : siemens
dgi/dt = -gi*(1./taui) : siemens
alpha_m = 0.32*(mV**-1)*(13*mV-v+VT)/
         (exp((13*mV-v+VT)/(4*mV))-1.)/ms : Hz
beta_m = 0.28*(mV**-1)*(v-VT-40*mV)/
        (exp((v-VT-40*mV)/(5*mV))-1)/ms : Hz
alpha_h = 0.128*exp((17*mV-v+VT)/(18*mV))/ms : Hz
beta_h = 4./(1+exp((40*mV-v+VT)/(5*mV)))/ms : Hz
alpha_n = 0.032*(mV**-1)*(15*mV-v+VT)/
         (exp((15*mV-v+VT)/(5*mV))-1.)/ms : Hz
beta_n = .5*exp((10*mV-v+VT)/(40*mV))/ms : Hz
zetav = zeta(v/mV):1
'''
```

I also tried (constant over dt) flag but it gives the same NotImplementedError. Thanks!

**Comment by mstimberg** [comment]
Hi, you get a `NotImplementedError` when you try to actually use the function (the same would happen if you tried to record it, for example). As I explained in my previous comment, you only provide a Python implementation and can therefore only execute your function with Python. By default, Brian generates code for Cython (or C++ via weave if you are using Python 2), and it cannot use your function with this code. You can switch to Python code generation globally with `prefs.codegen.target = 'numpy'`, but this reduces the speed of your simulation. Alternatively you can provide a C++/Cython implementation of your function as shown in the documentation (see link in my previous comment). Or, as I would suggest, don't use a function but define `zetav` via the equations.

**Comment by mstimberg** [comment]
I'm closing this issue for now, please feel free to re-open it if you still have problems with this.
